### ISSUE TITLE
external.Get took a long time

### ISSUE THREAD

**Issue by xrmzju** [kubernetes-sigs/cluster-api#1663]
**What steps did you take and what happened:**
i create about 60k machines in my env, but everytime cluster-api restart or relist, it tooks a long time to process the whole item, the bottleneck is machinecontroller from metrics record by prometheus
![work_duration](https://user-images.githubusercontent.com/27884109/67654446-960fd800-f988-11e9-8cb8-ebd3e9805d32.png)
![workqueue_depth](https://user-images.githubusercontent.com/27884109/67654454-9ad48c00-f988-11e9-998e-feb02e714a1c.png)

functions that took long times were dig out through add extra logs in machine controller: mainly because of external.Get func. it spend about 5 seconds every time we call it(reconcile Bootstrap and reconcileInfraStructure). client.Get() uses `unstructuredClient` to get unstructured obj, which will use `dynamicResourceClient` and send request to apiserver in every `Get` action. the request speed is limited by the QPS of the client(default is 20).  i scaled the qps to 200, and `client.Get()` took about 0.5s.

**Anything else you would like to add:**
since we have list-watch  external obj in cluster-api, i think external.Get func should take advantage of it and Get from the cache. it will also reduce the  num of requests sent to apiserver

/kind bug

**Comment by xrmzju** [comment]
is there a way to finish this?

**Comment by xrmzju** [comment]
> is there a way to finish this?

the reason is machine controller uses [DelegatingReader](https://github.com/kubernetes-sigs/controller-runtime/blob/ad57a976ffa1eac39bc9d9cb267ca5489fb8f038/pkg/manager/manager.go#L325),who Get and List requests for unstructured types using ClientReader(request directly to apiserver) while requests for any other type of object with use the CacheReader。i fix this by using  [NewClientFunc](https://github.com/kubernetes-sigs/controller-runtime/blob/ad57a976ffa1eac39bc9d9cb267ca5489fb8f038/pkg/manager/manager.go#L182) options when calling `manager.New`。 the `newClientFunc` returns a client uses cacheReader to do get and list, while clientReader to do write action.
`func newClientFunc(cache cache.Cache, config *rest.Config, options client.Options) (client.Client, error) {
	// Create the Client for Write operations.
	c, err := client.New(config, options)
	if err != nil {
		return nil, err
	}

	return &client.DelegatingClient{
		Reader:       cache,
		Writer:       c,
		StatusClient: c,
	}, nil
}
`

**Comment by ncdc** [comment]
I was about to paste the same thing, from https://github.com/kubernetes-sigs/controller-runtime/issues/615#issuecomment-536550773 😄

**Comment by ncdc** [comment]
In the linked controller-runtime issue, they say they don't use the cache by default for unstructured objects out of concern that you might accidentally/unknowingly end up caching all the resources in a cluster, and so it's up to someone (i.e., us) to override that. Given that we're already starting shared informers for all our `external` types, we should take advantage of the cache. I'm 👍 to making this change.

**Comment by xrmzju** [comment]
> In the linked controller-runtime issue, they say they don't use the cache by default for unstructured objects out of concern that you might accidentally/unknowingly end up caching all the resources in a cluster, and so it's up to someone (i.e., us) to override that. Given that we're already starting shared informers for all our `external` types, we should take advantage of the cache. I'm 👍 to making this change.

nice.. i can't believe i did not find the issue, i have been struggling whole day to fix this...

**Comment by ncdc** [comment]
@detiber @vincepri @chuckha @liztio WDYT?

**Comment by vincepri** [comment]
I haven't read the whole linked issue yet, but how do we prevent caching all resources in a cluster?

**Comment by ncdc** [comment]
Every reconciler we create results in caching all the resources of that specific type (e.g. the `ClusterReconciler` results in a shared informer that caches all `Cluster` resources). Additionally, any `.Watches()` calls that we make with a specific `source.Kind` result in a cache of that kind. So we're already caching all Clusters, MachineDeployments, MachineSets, Machines. And when we `reconcileExternal` to watch the infra & bootstrap resources, we're caching those too.

In other words, we will currently cache every **typed** resource we use the `Manager's` `Client` for, plus everything **unstructured** we use `reconcileExternal` for.

The thing that we're missing is using the cache against unstructured types.

Do we have any code that reads unstructured objects that we don't want cached?

**Comment by xrmzju** [comment]
> I haven't read the whole linked issue yet, but how do we prevent caching all resources in a cluster?

`InformersMap` only start a informer for specific gvk  when the `Get` func  invoked

**Comment by xrmzju** [comment]
any idea to cache remote Cluster's nodes && secrets also? we list nodes too frequently, it is really expensive @vincepri @ncdc @detiber

**Comment by xrmzju** [comment]
> any idea to cache remote Cluster's nodes && secrets also? we list nodes too frequently, it is really expensive @vincepri @ncdc @detiber

we list nodes for every reconcile. https://github.com/kubernetes-sigs/cluster-api/blob/f7e048646ae89c6e0ddf755c3760f233017386d9/controllers/machine_controller_noderef.go#L103

**Comment by ncdc** [comment]
Let's file a separate issue for that?

**Comment by detiber** [comment]
I like the idea of adding the caching for unstructured resources. If for some reason we end up needing to avoid caching a specific resource type in the future I think we can cross that bridge when we get there.
